linera-core/src/client/mod.rs

```diff
@@ -1203,143 +1159,144 @@ where
             .known_committees()
             .await
             .map_err(|_| NodeError::InvalidChainInfoResponse)?;
+
         // Retrieve newly received certificates from this validator.
         let query = ChainInfoQuery::new(chain_id).with_received_log_excluding_first_n(tracker);
         let info = remote_node.handle_chain_info_query(query).await?;
-        let mut tracker = MultichainTracker::new(tracker);
         let mut other_sender_chains = Vec::new();
-        let remote_log = info.requested_received_log.clone();
-        let remote_node_chains_view = info.requested_received_log.into_iter().fold(
-            BTreeMap::<ChainId, Vec<BlockHeight>>::new(),
+        let remote_log = info.requested_received_log;
+        let remote_node_chains_view = remote_log.iter().fold(
+            BTreeMap::<ChainId, BlockHeight>::new(),
             |mut chain_to_info, entry| {
                 chain_to_info
                     .entry(entry.chain_id)
-                    .or_default()
-                    .push(entry.height);
+                    .and_modify(|h| *h = entry.height.max(*h))
+                    .or_insert(entry.height);
                 chain_to_info
             },
         );
 
-        // Collect all certificates that have been successfully processed.
-        let mut certificates: Vec<Certificate> = Vec::new();
-
-        for (chain_id, mut block_batch) in remote_node_chains_view {
-            block_batch.sort();
+        // Obtain the chain heights we already have from the local node.
+        let futures = remote_node_chains_view
+            .keys()
+            .map(|sender_chain_id| async move {
+                let query = ChainInfoQuery::new(*sender_chain_id);
+                let local_response = self.local_query(query).await?;
+                Ok::<_, NodeError>(local_response.info.next_block_height)
+            })
+            .collect::<Vec<_>>();
+        let local_next_heights = stream::iter(futures)
+            .buffer_unordered(chain_worker_limit)
+            .try_collect::<Vec<_>>()
+            .await?;
```

Comment:
This can be extracted to a function which takes as an argument an iterator over chain IDs and returns `local_chain_highest`.